source/common/protobuf/utility.h

```diff
@@ -177,6 +178,51 @@ class MessageUtil {
   static void loadFromYaml(const std::string& yaml, Protobuf::Message& message);
   static void loadFromFile(const std::string& path, Protobuf::Message& message);
 
+  static void checkForDeprecation(const Protobuf::Message& message, bool fatal) {
+    const Protobuf::Descriptor* descriptor = message.GetDescriptor();
+    const Protobuf::Reflection* reflection = message.GetReflection();
+    for (int i = 0; i < descriptor->field_count(); ++i) {
+      const auto* field = descriptor->field(i);
+
+      // If this field is not in use, continue.
+      if ((field->is_repeated() && reflection->FieldSize(message, field) == 0) ||
+          (!field->is_repeated() && !reflection->HasField(message, field))) {
+        continue;
+      }
+
+      // If this field is deprecated, warn or throw an error.
+      if (field->options().deprecated()) {
+        std::string err = fmt::format(
+            "Using deprecated option '{}'. This configuration will be removed from Envoy soon. "
+            "Please see https://github.com/envoyproxy/envoy/blob/master/DEPRECATED.md for "
+            "details.",
+            field->name());
+        if (fatal) {
+          throw ProtoValidationException(err, message);
+        } else {
+          ENVOY_LOG_MISC(warn, "{}", err);
```

Comment:
use config logger?

Reply:
Which logger?  The only other log macro in the util.cc used MISC so I didn't think any of the other log macros were suitable.

Reply:
We could wrap in a class to provide access to the config logger (or maybe there is a more direct way?).

Reply:
Just make MessageUtil inherit Loggable<config> and it should be good (it only provides static method)

Reply:
Ah, so you want it to inherit from Logger::Loggable<Logger::Id::config> 

Done, though I think ENVOY_LOG_MISC might make more sense, given that someone could put non-config related proto utilities in source/common/protobuf/utility.h 

That said the main uses of proto are config and gRPC which has its own directories, so it's fine for now.